File: onda_freq.py
# Calcula parametros de onda no dominio da frequencia
#
# Elaborado por Henrique P. P. Pereira (henriqueppp@peno.coppe.ufrj.br)
#
# Ultima modificacao: 01/11/2012
#
# Dados de entrada: dt - intervalo de amostragem
#                   h - profundidade
#                   eta - vetor de elevacao
#                   etax - vetor de deslocamento em x
#                   etay - vetor de deslocamento em y
#
# Dados de saida: f - vetor de frequencias
#                 an - auto-espectro de eta
#                 anx - auto-espectro de etax
#                 any - auto-espectro de etay
#                 a1, b1 - coeficientes de fourier de 1 ordem
#                 diraz - vetor de direcao azimutal em graus
#                 dirm - vetor de direcao media
#                 dirtp - direcao associado a frequencia de pico
#                 fp - frequencia de pico
#                 tp - periodo de pico
#                 hm0 - altura significativa
#
# Subrotinas chamadas: spec
#                      spec2
#                      numeronda

##Importa bibliotecas
from numpy import *
from pylab import find
from espectro import espec1, espec2
from numeronda import numeronda

##carrega arquivo de onda (pc henrique)
#[t,eta,dspx,dspy]=loadtxt('/home/henrique/Dropbox/Tese_Mestrado/consistencia_python/arq_200905010000.txt', unpack=True)

##carrega arquivo de onda (coppe)
[t,eta,dspy,dspx] = loadtxt('/home/hppp/Dropbox/Tese_Mestrado/consistencia_python/arq_200905010000.txt', unpack=True)



#comprime o tamanho do vetor para potencia de 2
t = t[0:1024]
eta = eta[0:1024]
dspy = dspy[0:1024]
dspx = dspx[0:1024]

#profundidade
h=500

#intervalo de tempo
dt = t[1]-t[0]

#janela
han = 0

#graus de liberdade
gl = 32

#calculo do espectro simples da elevacao
aan = espec1(eta,dt,gl,han)

#calculo dos espectros cruzados da elevacao e deslocamentos horizontais
aannx = espec2(eta,dspx,dt,gl,han)
aanny = espec2(eta,dspy,dt,gl,han)

#vetor de frequencia
f = aan[:,0]

#calculo do numero de onda
k = numeronda(h,f,len(f))

#definicao dos parametros dos espectros

#auto-espectro de eta
sn = aan[:,1]

#auto-espectro de dsp x
snx = aannx[:,2]

#auto-espectro de dsp y
sny = aanny[:,2]

#quad-espectro de eta e dsp x
qnnx = aannx[:,5]

#quad-espectro de eta e dsp y
qnny = aanny[:,5]

#calculo da direcao de onda

a1 = []
b1 = []
dirr = []

for i in range(len(f)):

	#calculo dos coeficientes de fourier
	a1.append( qnnx[i] / (k[i]*pi*sn[i]) )
	b1.append( qnny[i] / (k[i]*pi*sn[i]) )

	#cria numero imaginario com os coeficientes
	imag = complex(a1[i],b1[i])

	#calcula a direcao em radianos
	dirr.append(angle(imag))

#vetor de diracao para array
dirr = array(dirr)

#passa direcao de rad para graus
dirg = dirr * 180 / pi

#passa de trigonometrico para azimute
diraz = 270 - dirg

#muda de onde a onda vai para onde ela vem
#diraz = diraz - 180

#condicao para valores maiores que 360 e menores que 0

for i in range(len(diraz)):

	if diraz[i] > 360:

		diraz[i] = diraz[i]-360

	elif diraz[i] < 0:

		diraz[i] = diraz[i] + 360

#calcula parametros de onda

#periodo de pico
tp = float(1./(f[find(sn == max(sn))]))

#frequencia de pico
fp = 1/tp

#altura significativa (Hm0)

#variacao de frequencia
df = f[1] - f[0]

#momento de espectral m0
m0 = sum(sn)*df

#calculo da altura significativa
hm0 = 4.01*sqrt(m0)

#direcao do periodo de pico
dirtp = float(diraz[find(f==fp)])

File: onda_freq.m
% Calcula parametros de onda no dominio da frequencia
%
% Elaborado por Henrique P. P. Pereira (henriqueppp@peno.coppe.ufrj.br)
%
% Ultima modificacao: 01/11/2012
%
% Dados de entrada: dt - intervalo de amostragem
%                   h - profundidade
%                   eta - vetor de elevacao
%                   etax - vetor de deslocamento em x
%                   etay - vetor de deslocamento em y
%
% Dados de saida: f - vetor de frequencias
%                 an - auto-espectro de eta
%                 anx - auto-espectro de etax
%                 any - auto-espectro de etay
%                 a1, b1 - coeficientes de fourier de 1 ordem
%                 diraz - vetor de direcao azimutal em graus
%                 dirm - vetor de direcao media
%                 dirtp - direcao associado a frequencia de pico
%                 fp - frequencia de pico
%                 tp - periodo de pico
%                 hm0 - altura significativa
%
% Subrotinas chamadas: spec
%                      spec2
%                      numeronda
%


function [f,an,anx,any,a1,b1,diraz,dirm,dirtp,fp,tp,hm0]=onda_freq(dt,h,eta,etax,etay,gl)


%% Calculo do espectro pela funcao spectrum (subrotina spec e spec2)

%graus de liberdade
% gl=2;

%calculo do espectro
[aan]=spec(eta,dt,gl);
[aannx]=spec2(eta,etax,dt,gl);
[aanny]=spec2(eta,etay,dt,gl);

f=aan(:,1);

%% Calculo do numero de onda 'k' (subrotina numeronda) verificar com parente
[k]=numeronda(h,f,length(f))'; 

%% Calcula os quad-espectros necessarios para o calculo de a1 e b1
qnnx=aannx(:,5); %quad-espectro de eta e etax
qnny=aanny(:,5); %quad-espectro de eta e etay
an=aan(:,2) ;%auto-espectro de eta
anx=aannx(:,3); %auto-espectro de etax
any=aanny(:,3); %auto-espectro de etay

%% Calculo da direcao da onda

% Calculo de a1 e b1 em radianos
for i=1:length(f)
    
    a1(i,1)=qnnx(i,1)/(k(i,1)*pi*an(i,1)); % qual usar, com ou sem o autoespectro?
    
    b1(i,1)=qnny(i,1)/(k(i,1)*pi*an(i,1));
            
end

% Calcula o angulo em radianos (forma correto neste caso) utilizando funcao 'angle'
%dirr=angle(a1+b1.*j);

% Calcula da direcao pelo arcotangente em radianos
dirr=atan(b1./a1);

% Passa de rad. para grau (trigonometrico)
dirt=dirr.*(180/pi);

%corrige erro de nan que esta dando  de nan no ultimo elemento
dirt(end) = dirt(end - 1);

diraz = dirt;
% Passa de trigonometrico para azimute
diraz = trig_para_azim(dirt);

%diraz=270-dirt;

% Muda de onde a onda vai para onde a onda vem (confirmar com parente)
%diraz=diraz-180;

% Condicao para valores maior que 360 e menores que 0
for i=1:length(diraz)
    
    if diraz(i)>360;
        
        diraz(i)=diraz(i)-360;
        
    elseif diraz(i)<0
        
        diraz(i)=diraz(i)+360;
        
    end
end
  
% Correcao da declinacao magnetica
%nao foi feita mais a direcao esta certa. 
%provavelmente a bussola da boia ja deve corrigir

%% Calcula parametros de onda

% Periodo de pico no espectro de eta
fp=aan(find(aan(:,2)==max(aan(:,2))),1);
tp=1/fp;

% Altura Hm0 calculado pela espectro de eta

% Delta de frequencia
df=f(2)-f(1);

% Momento espectral de ordem zero
m0=0;

for i=1:length(an)
    
    m0=m0+(an(i,1)*df);
    
end

% Calculo da alt sig, utilizando momento espectral de ordem zero
hm0=4.01*sqrt(m0);

% Cria matriz para [f diraz] para o calculo da direco de pico
fdir=[f,diraz];

% Direcao associada ao periodo de pico
dirtp=fdir(find(fdir(:,1)==fp),2);

% Direcao media
dirm=mean(diraz);

% Acha o numero de onda associado a frequencia de pico

% Cria matriz para [f k] para o calculo do numero de onda associado a frequencia de pico
fk=[f,k];

%numero de onda associado a frequencia de pico
kfp=fk(find(fk(:,1)==fp),2);